### PR TITLE
treewide: fold -> foldr

### PR DESCRIPTION
**Pull request by siraben**
###### Motivation for this change
@Profpatsch continue with the deprecation?

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: lib/tests/misc.nix [hunk 1]
```diff
@@ -245,22 +245,20 @@ runTests {
...
+      # foldr with associative operator
       assoc = f builtins.add;
-      # fold with non-associative operator
+      # foldr with non-associative operator
       nonAssoc = f builtins.sub;
     in {
       expr = {
         assocRight = assoc foldr;
-        # right fold with assoc operator is same as left fold
+        # right foldr with assoc operator is same as left fold
```
Comment (SuperSandro2000) [inline lib/tests/misc.nix:256]: That comment is not right.

Comment (siraben) [reply lib/tests/misc.nix]: Thanks.

**Comment by SuperSandro2000** [comment]
I think you missed a file and ofborg complains about that.

**Comment by Profpatsch** [comment]
The question here is whether `foldr` is the right use anyway. Usually you really want `foldl'`, always, if you intend to use the result(s) of the fold.

**Comment by Profpatsch** [comment]
`foldr` is only useful if you depend on the laziness property, and idk how much that even applies in nix, since lists are forcing their elements to whnf.

**Comment by siraben** [comment]
Ok so should I go through my changes and make them `foldl'` when the function is associative and the element an identity element? e.g.

```diff
-    fold and true (attrValues (zipAttrsWithNames (attrNames pattern) (n: values:
+    foldr and true (attrValues (zipAttrsWithNames (attrNames pattern) (n: values:
```

might as well be

```diff
-    fold and true (attrValues (zipAttrsWithNames (attrNames pattern) (n: values:
+    foldl' and true (attrValues (zipAttrsWithNames (attrNames pattern) (n: values:
```

**Comment by Profpatsch** [comment]
` foldl' and true`

I’d argue that should use `lib.all` instead.

**Comment by siraben** [comment]
> ` foldl' and true`
> 
> I’d argue that should use `lib.all` instead.

Looks like we need an hlint equivalent for Nix! 😆

**Comment by Profpatsch** [comment]
but anyway, the refactor shouldn’t introduce any behaviour changes

**Review by Profpatsch — APPROVED** [review]
LGTM

Next up, somebody should go through uses of `foldr` to check whether we can gain memory or speed savings by using `foldl'` instead.

**Comment by Profpatsch** [comment]
cc @Infinisil @grahamc there might be easy memory/speed gains here

File: lib/default.nix [hunk 2]
```diff
@@ -79,7 +79,7 @@ let
       recursiveUpdate matchAttrs overrideExisting getOutput getBin
       getLib getDev getMan chooseDevOutputs zipWithNames zip
       recurseIntoAttrs dontRecurseIntoAttrs;
-    inherit (self.lists) singleton forEach foldr fold foldl foldl' imap0 imap1
+    inherit (self.lists) singleton forEach fold foldr foldl foldl' imap0 imap1
```
Comment (SuperSandro2000) [inline lib/default.nix:82]: Is this supposed to just change the position of foldr?

Comment (siraben) [reply lib/default.nix:82]: Ah, good catch, I'll revert this.

**Comment by tomberek** [comment]
running: `NIX_SHOW_STATS=1 time nix-instantiate nixos/release-combined.nix -A nixos.tests.misc.x86_64-linux --dry-run`

<details>
<summary>before</summary>

```
{
  "cpuTime": 4.4831,
  "envs": {
    "number": 8040823,
    "elements": 9909621,
    "bytes": 207930136
  },
  "list": {
    "elements": 5932327,
    "bytes": 47458616,
    "concats": 193616
  },
  "values": {
    "number": 12971839,
    "bytes": 311324136
  },
  "symbols": {
    "number": 118285,
    "bytes": 4366233
  },
  "sets": {
    "number": 1158225,
    "bytes": 305840304,
    "elements": 12357271
  },
  "sizes": {
    "Env": 16,
    "Value": 24,
    "Bindings": 8,
    "Attr": 24
  },
  "nrOpUpdates": 441021,
  "nrOpUpdateValuesCopied": 8904098,
  "nrThunks": 8005723,
  "nrAvoided": 10451397,
  "nrLookups": 3357854,
  "nrPrimOpCalls": 4350366,
  "nrFunctionCalls": 7254051,
  "gc": {
    "heapSize": 621469696,
    "totalBytes": 1090436352
  }
}
```
</details>

<details>
<summary>after</summary>

```
{
  "cpuTime": 4.45351,
  "envs": {
    "number": 8032348,
    "elements": 9892932,
    "bytes": 207661024
  },
  "list": {
    "elements": 5927432,
    "bytes": 47419456,
    "concats": 192848
  },
  "values": {
    "number": 12930424,
    "bytes": 310330176
  },
  "symbols": {
    "number": 118268,
    "bytes": 4429547
  },
  "sets": {
    "number": 1155028,
    "bytes": 300472592,
    "elements": 12134682
  },
  "sizes": {
    "Env": 16,
    "Value": 24,
    "Bindings": 8,
    "Attr": 24
  },
  "nrOpUpdates": 439018,
  "nrOpUpdateValuesCopied": 8709175,
  "nrThunks": 7971489,
  "nrAvoided": 10436936,
  "nrLookups": 3348099,
  "nrPrimOpCalls": 4344621,
  "nrFunctionCalls": 7246543,
  "gc": {
    "heapSize": 604692480,
    "totalBytes": 1083838032
  }
}
```
</details>

Most of the numbers improve <1%, and heapSize is reduced ~3%.

**Comment by siraben** [comment]
@tomberek that seems as expected, since the only gain here is saving the indirection from fold to foldr (as `fold = foldr;` in lists.nix).

What might be reasonable is going through the replacements and seeing where things definitely would be equivalent under `foldl` (that is, when the binary operation and the initial element form a monoid).

**Comment by siraben** [comment]
> Next up, somebody should go through uses of `foldr` to check whether we can gain memory or speed savings by using `foldl'` instead.

perhaps should be done in a separate PR as suggested by @Profpatsch

**Pull request merged**